File: cuda-cupy/dot.py
from math import sin, cos

from numpy import array
import cupy

from _dot import dot

blocks = 32
threads = 256
n = 10000

# initialise data and calculate reference values on CPU
x = array([sin(i) * 2.3 for i in range(n)], float)
y = array([cos(i) * 1.1 for i in range(n)], float)
z_ref = sum(x * y)

# allocate + copy initial values
x_ = cupy.asarray(x)
y_ = cupy.asarray(y)
z_ = cupy.empty(1, float)

# calculate dot product on GPU (and reduce partial sums)
dot(blocks, threads, n, x_.data.ptr, y_.data.ptr, z_.data.ptr)

# copy result back to host and print with reference
z = z_.get()[0]
print(' reference: {0}'.format(z_ref))
print('    result: {0}'.format(z))

File: hipblas/dot.c
#include <stdio.h>
#include <math.h>
#include <hip/hip_runtime.h>
#include <hipblas.h>

int main(void)
{
    int i;
    const int n = 100;
    double x[n], y[n];
    double *x_, *y_, *z_;
    double z_ref = 0.0;
    double z;
    hipblasHandle_t handle;

    // initialise data and calculate reference value on CPU
    for (i=0; i < n; i++) {
        x[i] = sin(i) * 2.3;
        y[i] = cos(i) * 1.1;
        z_ref += x[i] * y[i];
    }
    hipblasCreate(&handle);

    // use GPU pointers to keep data on GPU
    hipblasSetPointerMode(handle, HIPBLAS_POINTER_MODE_DEVICE);

    // allocate + copy initial values
    hipMalloc((void **)(&x_), sizeof(double) * n);
    hipMalloc((void **)(&y_), sizeof(double) * n);
    hipMalloc((void **)(&z_), sizeof(double));
    hipMemcpy(x_, x, sizeof(double) * n, hipMemcpyHostToDevice);
    hipMemcpy(y_, y, sizeof(double) * n, hipMemcpyHostToDevice);

    // calculate dot on GPU
    hipblasDdot(handle, n, x_, 1, y_, 1, z_);

    // copy result back to host and print with reference
    hipMemcpy(&z, z_, sizeof(double), hipMemcpyDeviceToHost);
    printf("reference: %f\n   result: %f\n", z_ref, z);

    hipblasDestroy(handle);
    return 0;
}
